### PR TITLE
slight refactor in `unexpand` ; fix and optimize `factor` ; fix `touch`, `test`, `tac`, `shuf`, `sleep` ; merge PR from @ctjhoa for `cksum`

### PR DESCRIPTION
**Pull request by kwantam**
This PR does the following:

1. a slight cleanup in `unexpand` that gets rid of redundant logic.
2. cherry-pick @wikol's PR #531 which adds Pollard's rho implementation to `factor`
3. Adds fast divisibility check and accompanying tables in `factor`, along with a generator for those tables.
4. Adds special cased arithmetic functions in `factor` to prevent overflow when input is larger than 2^63. This means we can always use Pollard's rho rather than falling back to trial division.
5. Adds three tests for `factor`.
6. Fixes `touch`.
  * Note that this utility does not work correctly when the `-h` option is supplied, as it uses Rust's `set_file_times()` stdlib function, which follows symlinks.
  * Fixing this requires using `lutimes`, but according to the man page that is only available on BSDs and Linux, so this will require a platform-specific solution.
7. Fixes `test`.
  * Parens remain unimplemented.
8. Fixes `tac`
9. Fixes `shuf`
10. Fixes `sleep`
11. Pulls in @ctjhoa's `cksum` PR from kwantam/coreutils#1

Closes #531 via cherry-pick.

### CODE REVIEW

File: src/factor/factor.rs [hunk 1]
```diff
@@ -12,66 +16,174 @@
...
+    let mut x = range.ind_sample(&mut rng);
+    let mut y = x;
+    let mut a = range.ind_sample(&mut rng);
+    let mut b = range.ind_sample(&mut rng);
+
+    loop {
+        x = rho_pollard_pseudorandom_function(x, a, b, num);
+        y = rho_pollard_pseudorandom_function(y, a, b, num);
+        y = rho_pollard_pseudorandom_function(y, a, b, num);
+        let d = gcd(num, max(x,y) - min(x,y));
```
Comment (Arcterus) [inline src/factor/factor.rs]: I'd prefer that there be spaces after the commas here.

File: src/factor/factor.rs [hunk 2]
```diff
@@ -12,66 +16,174 @@
...
+}
 
+fn rho_pollard_factor(num: u64, factors: &mut Vec<u64>) {
+    if is_prime(num) {
+        factors.push(num);
+        return;
+    }
+    let divisor = rho_pollard_find_divisor(num);
+    rho_pollard_factor(divisor, factors);
+    rho_pollard_factor(num/divisor, factors);
```
Comment (Arcterus) [inline src/factor/factor.rs]: Space between `num` and `/` and `/` and `divisor`.

File: src/factor/gen_table.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+#[cfg_attr(test, allow(dead_code))]
+fn main() {
+    // By default, we print the multiplicative inverses mod 2^64 of the first 10k primes
+    let n = args().skip(1).next().unwrap_or("10000".to_string()).parse::<usize>().ok().unwrap_or(10000);
+
+    print!("{}", PREAMBLE);
+
+    let m = n;
+    Sieve::new()
+        .scan((0,3), move |st,x| {
```
Comment (Arcterus) [inline src/factor/gen_table.rs]: Space between `,` and `x`.

File: src/factor/gen_table.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            let (count, mut cols) = *st;
+            if count < m {
+                // format the table
+                let outstr = format!("({},{},{}),", x, inv_mod_u64(x).unwrap(), MAX_U64/x);
```
Comment (Arcterus) [inline src/factor/gen_table.rs]: Space between `MAX_U64` and `/`, and `/` and `x`.

File: src/factor/gen_table.rs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            } else if count == m {
+                // now we're done formatting the table, print NEXT_PRIME
+                print!("\n];\n\npub const NEXT_PRIME: u64 = {};\n", x);
+
+                *st = (count + 1, cols);
+                Some(1)
+            } else {
+                None
+            }
+        }).take(m+1).count();
```
Comment (Arcterus) [inline src/factor/gen_table.rs]: Space between `m` and `+`, and `+` and `1`.

File: src/factor/factor.rs (hunk 2 above)
Comment (Arcterus) [inline src/factor/factor.rs]: There are a few more spots like this that I haven't commented on.  Can you fix those as well?

File: src/factor/factor.rs [hunk 6]
```diff
@@ -12,66 +16,174 @@
...
 extern crate getopts;
 extern crate libc;
+extern crate rand;
 
-use std::vec::Vec;
-use std::old_io::BufferedReader;
-use std::old_io::stdio::stdin_raw;
+use numeric::*;
+use prime_table::{P_INVS_U64, NEXT_PRIME};
+use std::cmp::{max,min};
```
Comment (Arcterus) [inline src/factor/factor.rs]: Space after comma.

**Comment by Arcterus** [comment]
Everything looks great other than the minor spacing issues. :)

**Comment by kwantam** [comment]
@Arcterus thanks for the review. I've taken care of the issues you pointed out, and I'm pretty sure I found the rest of them as well.

File: src/factor/factor.rs [hunk 7]
```diff
@@ -12,66 +16,174 @@
...
         num /= 2;
-        ret.push(2);
+        factors.push(2);
     }
-    let mut i = 3;
-    while i * i <= num {
-        while num % i == 0 {
-            num /= i;
-            ret.push(i);
+    if is_prime(num) {
```
Comment (wikol) [inline src/factor/factor.rs]: This might fail when num >= 2^63 as is_prime needs exponentiation modulo num.

Comment (kwantam) [inline src/factor/factor.rs]: > This might fail when num >= 2^63 as is_prime needs exponentiation modulo num.

I'll take a look into this. Thanks.

**Comment by Heather** [comment]
@kwantam sadly this should be rebased again :(

**Comment by kwantam** [comment]
@Heather no problem. Let me take a look at the potential issue @wikol pointed out above and then I'll rebase it and post another message to the effect that it's ready.

**Comment by kwantam** [comment]
@Heather I have addressed @wikol's concern regarding overflow, which incidentally makes the worst case performance of `factor` substantially better because we can now always use Pollard's rho method. Also added a couple tests to exercise possible overflow paths. This is ready for a merge.

**Comment by Heather** [comment]
Thank you, looks good to me.

**Pull request merged**